source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+    read_write_vector_index_set.fill_index_vector(out_deal);
+    std::vector<PetscInt> out_petsc(out_deal.begin(), out_deal.end());
+
+    std::vector<PetscInt> dummy;
+
+    this->do_reinit(in_petsc, dummy, out_petsc, dummy, communicator);
+  }
+
+  void
+  CommunicationPattern::reinit(
+    const std::vector<types::global_dof_index> &indices_has,
+    const std::vector<types::global_dof_index> &indices_want,
+    const MPI_Comm &                            communicator)
+  {
+    // Clean vectors from numbers::invalid_dof_index (indicating padding)
+    std::vector<PetscInt> indices_has_clean, indices_has_loc;
+    std::vector<PetscInt> indices_want_clean, indices_want_loc;
+    indices_want_clean.reserve(indices_want.size());
+    indices_want_loc.reserve(indices_want.size());
+    indices_has_clean.reserve(indices_has.size());
+    indices_has_loc.reserve(indices_has.size());
+
+    PetscInt loc         = 0;
+    bool     has_invalid = false;
+    for (const auto i : indices_has)
+      {
+        if (i != numbers::invalid_dof_index)
+          {
+            indices_has_clean.push_back(static_cast<PetscInt>(i));
+            indices_has_loc.push_back(loc);
+          }
+        else
+          has_invalid = true;
+        loc++;
+      }
+    if (!has_invalid)
+      indices_has_loc.clear();
+
+    loc         = 0;
+    has_invalid = false;
+    for (const auto i : indices_want)
+      {
+        if (i != numbers::invalid_dof_index)
+          {
+            indices_want_clean.push_back(static_cast<PetscInt>(i));
+            indices_want_loc.push_back(loc);
+          }
+        else
+          has_invalid = true;
+        loc++;
+      }
+    if (!has_invalid)
+      indices_want_loc.clear();
+
+    this->do_reinit(indices_has_clean,
+                    indices_has_loc,
+                    indices_want_clean,
+                    indices_want_loc,
+                    communicator);
+  }
+
+  void
+  CommunicationPattern::do_reinit(const std::vector<PetscInt> &inidx,
+                                  const std::vector<PetscInt> &inloc,
+                                  const std::vector<PetscInt> &outidx,
+                                  const std::vector<PetscInt> &outloc,
+                                  const MPI_Comm &             communicator)
+  {
+    clear();
+
+    // inidx is assumed to be unstructured and non-overlapping.
+    // However, it may have holes in it and not be a full cover.
+    //
+    // We create two PETSc SFs and compose them to get
+    // the final communication pattern
+    //
+    //  sf1 : local distributed to tmp
+    //  sf2 : tmp to local with ghosts
+    //  sf(x) = sf2(sf1(x))
+    PetscSF sf1, sf2;
+
+    // First create an SF where leaves are inidx (at location inloc)
+    // and roots are unique indices in contiguous way
+    // Code adapted from MatZeroRowsMapLocal_Private in PETSc
+    PetscInt n  = static_cast<PetscInt>(inidx.size()), nl;
```

Comment:
I read this wrong the first time around, believing that you're using the comma operator here. Let's do this instead:
```suggestion
    const PetscInt n  = static_cast<PetscInt>(inidx.size());
    PetscInt nl;
```